CIP-0015/CIP-0015.md

```diff
@@ -33,57 +33,88 @@ A voting key is simply an ED25519 key. How this key is created is up to the wall
 
 ### Associating stake with a voting key
 
+This method has been used since Fund 2.
+For future fund iterations, a new method making use of time-lock scripts may
+be introduced as described [below][future-development].
+
 Recall: Cardano uses the UTXO model so to completely associate a wallet's balance with a voting key (i.e. including enterprise addresses), we would need to associate every payment key to a voting key individually. Although there are attempts at this (see [CIP8](../CIP-0008/CIP-0008.md)), the resulting data structure is a little excessive for on-chain metadata (which we want to keep small)
 
 Given the above, we choose to only associate staking keys with voting keys. Since most Cardano wallets only use base addresses for Shelley wallet types, in most cases this should perfectly match the user's wallet.
 
 ### Registration metadata format
 
-A Catalyst registration transaction are a regular Cardano transaction with a specific transaction metadata associated with it
+A Catalyst registration transaction is a regular Cardano transaction with a specific transaction metadata associated with it.
 
-Notably, there should be three entries inside the metadata map:
+Notably, there should be four entries inside the metadata map:
 
-Voting key registration
+Voting registration example:
 ```
 61284: {
   // voting_key - CBOR byte array
   1: "0xa6a3c0447aeb9cc54cf6422ba32b294e5e1c3ef6d782f2acff4a70694c4d1663",
   // stake_pub - CBOR byte array
   2: "0xad4b948699193634a39dd56f779a2951a24779ad52aa7916f6912b8ec4702cee",
-  // address - CBOR byte array
-  3: "0x00588e8e1d18cba576a4d35758069fe94e53f638b6faf7c07b8abd2bc5c5cdee47b60edc7772855324c85033c638364214cbfc6627889f81c4"
+  // reward_address - CBOR byte array
+  3: "0x00588e8e1d18cba576a4d35758069fe94e53f638b6faf7c07b8abd2bc5c5cdee47b60edc7772855324c85033c638364214cbfc6627889f81c4",
+  // slot_number
+  4: 5479467
 }
 ```
 
-Signing the voting public key with the staking private key
+The entries under keys 1, 2, 3, and 4 represent the Catalyst voting key,
+the staking key on the Cardano network, the address to receive rewards,
+and the current slot number, respectively. A registration with these metadata
+will be considered valid if the following conditions hold:
+
+- The slot number of the block including this transaction is equal to
```

Comment:
agree, the user may need even several minutes to sign the transaction on a hardware wallet (which needs to know the slot number in advance). Could you @mzabaluev explain the rationale behind this constraint?

I gather that the motivation behind this constraint is preventing the users from submitting a key registration with too big value of the counter which would effectively make their key un-updatable - for that purpose a constraint that the slot number should be no higher than the slot number of the block in which the registration is submitted should do.

Another motivation could be avoiding state storage - Not sure how the current "active voting keys" are stored but storing one counter alongside them (to be able to compare it against subsequent re-registration) seems like more viable than forcing such strict consistency with the block number where the key registration would be submitted

Reply:
I agree that the two slot time window is too short, so either a larger validity timeout is needed or a different mechanism to provide a unique nonce.

> it would allow any pool to censor Catalyst registration transactions by purposely skipping them rendering them invalid for future slots.

What number of slots would be sufficient to prevent such censorship, assuming an honest majority among the pools?
This time window defines how far apart successive re-registrations need to be for the same staking key, to make it impossible to nullify the newer registration by replaying the older.

> I gather that the motivation behind this constraint is likely preventing the users from submitting a key registration with too big of a value of the counter which would effectively make their key un-updatable

It's rather so that old registrations cannot be replayed, reverting a possible newer registration of a different voting key (or a revocation, whenever that use case is supported) with the same staking key. But it could also prevent the problem you describe, if we apply the "larger slot number wins" rule to resolving conflicting registrations rather than the order of transactions (which may be problematic if the transactions happen to be in the same block and there is no UTxO-imposed order between them).

> Not sure how the current "active voting keys" are stored but they probably need to be stored somehow and storing a counter alongside them (to be able to compare it against subsequent re-registrations) seems more viable

It's certainly doable for the validation logic on the snapshotting tool, but I'm not sure about all wallets.
A wallet would be expected to recover the last successful registration transaction for the selected public staking key to obtain the value of the counter. If this is compatible with constraints posed by HW wallets, we can change the specification to use this stateful scheme.

Reply:
> A wallet would be expected to recover the last successful registration transaction for the selected public staking key to obtain the value of the counter.

I'm not sure why would a wallet need that - if the counter is chosen as the current slot number, that's a global increasing value. If we are afraid about the counter getting bigger than it should be, we can just ignore registrations with a value of the counter higher than the slot where they have been submitted. However, if a user somehow signs by mistake a registration with a future slot number, somebody may abuse that to overwrite their registration later on, but that's probably a problem that neither the currently proposed design addresses.

Reply:
slot number is only a suggestion of what to use for the nonce. It's nice because it always increases, but fundamentally, it can be any number that never goes backwards. e.g. 1, 2, 3, 4, 3 would result in 4 being the one used and ignore the last. Note that this nonce is not implemented for fund4 (it's included, but snapshot ignores it). We want to get the schema correct so we can properly scope out adding this for fund 5, but also want to stabilize so teams aren't having to continuously make changes.